### PR TITLE
Reuse values as far as possible

### PR DESCRIPTION
**Pull request by KeenS**
This patch is just for your information. You may or may not merge this patch.

Unlike functional programming languages, Rust's memory allocation is slow so allocating and copying should be avoided as far as possible, especially in simulation code, that uses the same values in loops.

* Stop deriving copy on > 64bit data
* Use references as far as possible

this patch let your program render a image in 2.47s -> 2.00s on my laptop, debug build.

The compiler and the borrow checker may help you to reduce allocations.

### CODE REVIEW

**Comment by krtx** [comment]
@KeenS 

Thank you for your patch.  I think that I should learn about Rust's memory allocation, and I should not have ignore the messages by the borrow checker...  I didn't mind about the efficiency of memory allocation when developing, so I am very happy to apply your patch.

I have one question: If I apply your patch, then when `c := a + b` is evaluated, it changes the value of `a`.  I think it contradicts to my intuition that the operator `+` should not change the values of the operands.  Is it acceptable in Rust?  Or, is it wrong to overload `+` in the way that `+` has side effect?

I am also interested in the following:

> Unlike functional programming languages, Rust's memory allocation is slow

Could you tell me literature that explain this?  Or, if this is explained in the Rust Book, could you tell me the section about this? (Yes, of course I will read entire the book, but it will be very helpful if you would tell me...)

**Comment by KeenS** [comment]
Yes, in rust, addition is just an operator and may mutate the addend. That way is adopted even in standard library (See [`Add` impl of `String`](https://github.com/rust-lang/rust/blob/master/src/libcollections/string.rs#L1629))

Memory allocation and performance is deeply related to the memory management. Most of the functional programming language adopts [Copy GC (ja)](http://seesaawiki.jp/w/author_nari/d/GC/standard/Copying) that featured by memory compaction, which makes allocation very fast. On the other hand, rust employs simple `malloc`/`free` to reduce runtime systems, and not to change the memory addresses of data. `Malloc` is slow because it needs to find free memory on each allocation.
In summary, because Rust has less control over memory and pointers *at runtime* , allocation in rust is slow.
